### ISSUE TITLE
How to center image in a window

### ISSUE THREAD

**Issue by aCuria** [ocornut/imgui#2212]
How do you center a single image in a window?
I tried "ImGui::PushItemWidth()" but it does not seem to do anything.

**Screenshots/Video** _(you can drag files here)_
![image](https://user-images.githubusercontent.com/35476457/49212190-68c74c80-f3fc-11e8-8709-c4e818df41b8.png)

**Comment by ocornut** [comment]
Use `SetCursorPos((GetWindowSize() - image_size) * 0.5f);` before the image call.

**Comment by aCuria** [comment]
I am currently doing the following:

`auto initialCursorPos = toVec2f(ImGui::GetCursorPos());`
`auto windowSize = toVec2f(ImGui::GetWindowSize());`
`auto centralizedCursorpos = (windowSize - textureDisplaySize_) * 0.5f;`
`centralizedCursorpos = clamp(centralizedCursorpos, initialCursorPos, centralizedCursorpos);`
`ImGui::SetCursorPos(toImVec2(centralizedCursorpos));`

The clamp() is to deal with the situation where the intended image is larger than the window area, this causes the centralizedCursorpos variable to become negative. When this happens, the scroll bars do not function as expected (unable to scroll through the entire image)

However, there is another issue that occurs when the titlebar is enabled
`ImGuiWindowFlags_NoTitleBar = false`

Here the centralization fails as it seems to be shifted up by the size of the title bar. I tried to use the following but it does not work as expected either:
`auto windowSize = toVec2f(ImGui::GetContentRegionAvail());`
`auto windowSize = toVec2f(ImGui::GetContentRegionMax());`

Whats the best way to fix this? I could not find a ImGui::getTitleBarHeight() function.

As a side note, its kinda annoying that ImVec2 is unable to do vector calculations - 
eg: `(windowSize - textureDisplaySize_) * 0.5f;`

**Comment by ocornut** [comment]
Please provide a clear repro.

> As a side note, its kinda annoying that ImVec2 is unable to do vector calculations -
eg: (windowSize - textureDisplaySize_) * 0.5f;

You are supposed to use your own maths types and use the imconfig.h facility to convert your maths types to ImVec2 You may also use/copy the operators from imgui_internal.h

**Comment by aCuria** [comment]
> Please provide a clear repro.

Using the code `SetCursorPos((GetWindowSize() - image_size) * 0.5f);`

With the title bar off, everything works as expected:
![image](https://user-images.githubusercontent.com/35476457/49978659-7e08b300-ff86-11e8-92f4-5532998be490.png)

With the title bar on, the image is slightly displaced towards the top:
![image](https://user-images.githubusercontent.com/35476457/49978592-371abd80-ff86-11e8-80b0-10141b833117.png)

Sometimes, the image can be slightly occluded (at the top) although there is sufficient space to display the entire image
![image](https://user-images.githubusercontent.com/35476457/49978900-a218c400-ff87-11e8-9a9c-4e53546d7e04.png)

I am guessing I need something like  `SetCursorPos((GetWindowSize() - TitleBarHeight() - image_size) * 0.5f); + TitleBarHeight()`. However, I cant find a "getTitleBarHeight()" function anywhere.

@ocornut Thanks for the help so far, ImGui is one of the best UI libraries I have come across

**Comment by aCuria** [comment]
The other "issue" I found (may not really be an issue) is that when the following code is used `(windowSize - textureDisplaySize_) * 0.5f;` It is possible for part of the image to be cropped off (unable to scroll up to see the top of the image. 

The fix I have is to do `centralizedCursorpos = clamp(centralizedCursorpos, initialCursorPos, centralizedCursorpos);` which will ensure that the new cursorPos is not smaller than the initialCursorPos. Leaving this here in case someone needs it


![image](https://user-images.githubusercontent.com/35476457/49979872-f9209800-ff8b-11e8-90ad-7413128976f5.png)

**Comment by aCuria** [comment]
I'm not sure if this should be a new issue or not, but I a different problem related to implementing this feature.

The problem I have occurs when calling the following code:

`ImGui::Image(texture, texturesize);`
`ImGui::SetScrollX(result.scroll_.x);`
`ImGui::SetScrollY(result.scroll_.y);`

When texturesize changes, the scrollbars are updated one frame late. This leads to the image "jumping" around. 

Calling SetScrollX before ImGui::Image() leads to the same behavior.

**Comment by demucin** [comment]
Hello
I'm not very familiar or experienced in ImGui.
This code is my starting point. 
<
        ImGui::Begin("Available Region");
        ImGui::GetWindowDrawList()->AddRect(
                    ImGui::GetCursorScreenPos(),
                    ImGui::GetCursorScreenPos() + ImGui::GetContentRegionAvail(), ImColor(255, 0, 0), 3.0);
        ImGui::End();
>

**Comment by aCuria** [comment]
Hi Demucin, 
I am not too sure how this is related to this topic? What are you trying to achieve with the code you listed?

**Comment by ocornut** [comment]
> With the title bar on, the image is slightly displaced towards the top: image

Your issue for that is that you should use `GetContentRegionAvail()` and not `GetWindowSize()`, which include decorations, etc.

> When texturesize changes, the scrollbars are updated one frame late. This leads to the image "jumping" around.

Haven't tried it so cannot answer, but you can now call `SetNextWindowScroll()` prior to `Begin()` to avoid a frame of lag when manipulating scroll programmatically

**Comment by zeyangl** [comment]
Hi, I'm hitting a similar issue, where the window size seems to account for scroll bar width, but the GetContentRegionAvail() always gives me the whole content size, not the visible window region size. What's the method of getting actual visible region, without deco or scroll bar?

**Comment by rokups** [comment]
To best understand what rect you need you should check out metrics window, accessible from demo > tools menu.

![image](https://user-images.githubusercontent.com/19151258/95548841-96c14300-0a0e-11eb-94fd-7f60c5ed7265.png)

You need `ImGui::GetCurrentWindow()->InnerRect` (include `imgui_internal.h`).

**Comment by mrhelmut** [comment]
Hello there!

Just dropping by to share that I'm running into the same centering oddity when title bars are involved.
```SetCursorPos((GetContentRegionAvail() - size) * 0.5f);``` won't exactly center within the available region. It seems that ```SetCursorPos()``` doesn't set the cursor position within the available region, but relatively to the full windows region (thus ignoring the presence of a title bar).

Is that the expected behavior of ```SetCursorPos()```?

![image](https://user-images.githubusercontent.com/3427451/115893756-d99db200-a458-11eb-8e5a-399c1c8afef2.png)

Cheers!

**Comment by jourrekk** [comment]
My recommendation is availmin = GetWindowContentRegionMin(); avail = GetWindowContentRegionMax() - availmin;
SetCurrentPos((avail - size)/2 + availmin);

**Comment by ocornut** [comment]
> Just dropping by to share that I'm running into the same centering oddity when title bars are involved.
SetCursorPos((GetContentRegionAvail() - size) * 0.5f); won't exactly center within the available region. It seems that SetCursorPos() doesn't set the cursor position within the available region, but relatively to the full windows region (thus ignoring the presence of a title bar).
> Is that the expected behavior of SetCursorPos()?

Sorry for late answer..

SetCursorPos() / GetCursorPos() are relative to window top-left corner. It's not great IHMO and nowadays would strongly suggest using SetCursorScreenPos() / GetCursorScreenPos() which are the "absolute" version.

Either way the issue with your code:
```
SetCursorPos((GetContentRegionAvail() - size) * 0.5f);
```
is that since `GetContentRegionAvail()` compute available space from the _current position_ (whatever that is), so the formula should:
```
SetCursorPos(GetCursorPos() + (GetContentRegionAvail() - size) * 0.5f);
```
Which is simpler than your solution.
Basically your line above imply that GetCursorPos() would return (0,0) after Begin().

**Comment by ocornut** [comment]
Closing this old topic as supposedly answered.
